substrate/runtime/contract/src/vm.rs

```diff
@@ -0,0 +1,785 @@
+	runtime: Runtime<T, E>,
+	run_err: Option<sandbox::Error>,
+) -> Result<ExecutionResult, Error> {
+	let mut return_data = Vec::new();
+
+	// Check the exact type of the error. It could be plain trap or
+	// special runtime trap the we must recognize.
+	match (run_err, runtime.special_trap) {
+		// No traps were generated. Proceed normally.
+		(None, None) => {}
+		// Special case. The trap was the result of the execution `return` host function.
+		(Some(sandbox::Error::Execution), Some(SpecialTrap::Return(rd))) => return_data = rd,
+		// Any other kind of a trap should result in a failure.
+		(Some(_), _) => {
+			return Err(Error::Invoke);
+		}
+		_ => {
+			// All possible cases should have been handled.
+			// If the control flow reached here, then it is a logic error.
+			unreachable!();
+		}
+	}
+
+	Ok(ExecutionResult {
+		return_data,
+	})
+}
+
+/// The result of execution of a smart-contract.
+#[derive(PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Debug))]
+pub struct ExecutionResult {
+	pub return_data: Vec<u8>,
+}
+
+impl ExecutionResult {
+	/// The result produced by the execution of the contract.
+	///
+	/// The contract can designate some buffer at the execution time via a special function.
+	/// If contract called this function with non-empty buffer it will be copied here.
+	///
+	/// Note that gas is already charged for returning the data.
+	pub fn return_data(&self) -> &[u8] {
+		&self.return_data
+	}
+}
+
+/// Execute the given code as a contract.
+pub fn execute<'a, T: Trait, E: Ext<T>>(
+	code: &[u8],
+	ext: &'a mut E,
+	gas_meter: &mut GasMeter<T>,
+) -> Result<ExecutionResult, Error> {
+	// TODO: Receive data as an argument
+
+	// ext_gas(amount: u32)
+	//
+	// Account for used gas. Traps if gas used is greater than gas limit.
+	//
+	// - amount: How much gas is used.
+	fn ext_gas<T: Trait, E: Ext<T>>(
+		e: &mut Runtime<T, E>,
+		args: &[sandbox::TypedValue],
+	) -> Result<sandbox::ReturnValue, sandbox::HostError> {
+		let amount = args[0].as_i32().unwrap() as u32;
+		let amount = <T::Gas as As<u32>>::sa(amount);
+
+		match e.gas_meter.charge(amount) {
+			GasMeterResult::Proceed => Ok(sandbox::ReturnValue::Unit),
+			GasMeterResult::OutOfGas => Err(sandbox::HostError),
+		}
+	}
+
+	// ext_put_storage(location_ptr: u32, value_non_null: u32, value_ptr: u32);
+	//
+	// Change the value at the given location in storage or remove it.
+	//
+	// - location_ptr: pointer into the linear
+	//   memory where the location of the requested value is placed.
+	// - value_non_null: if set to 0, then the entry
+	//   at the given location will be removed.
+	// - value_ptr: pointer into the linear memory
+	//   where the value to set is placed. If `value_non_null` is set to 0, then this parameter is ignored.
+	fn ext_set_storage<T: Trait, E: Ext<T>>(
+		e: &mut Runtime<T, E>,
+		args: &[sandbox::TypedValue],
+	) -> Result<sandbox::ReturnValue, sandbox::HostError> {
+		let location_ptr = args[0].as_i32().unwrap() as u32;
```

Comment:
same for this and the following lines.

Reply:
let's move this discussion in [this thread](https://github.com/paritytech/polkadot/pull/345#discussion_r203060428).